### PR TITLE
[BUGFIX] Fix server crash under rebuilding of RTEs and RelOptInfos

### PR DESCRIPTION
**Pull request by maksm90**
## Problem description

The GROUPING SETS statement with multiple canonical rollups working on randomly distributed partitioned table causes [rebuilding](https://github.com/greenplum-db/gpdb/blob/6.10.1/src/backend/optimizer/plan/plangroupext.c#L2231-L2292) of `root->simple_rte_array` and `root->simple_rel_array` on planning stage by PostgreSQL optimizer. The rebuilding process initializes both RTEs and RelOptInfos item by item and doesn't take into account that the `build_simple_rel()` routine is recursive for inherited relations and requires a full list of already initialized child RTEs inside `root->simple_rte_array`. As result the call of `build_simple_rel()` fails on the access to child RTEs.

The current fix segregates loops of building `root->simple_rte_array` and `root->simple_rel_array` trying to leave a whole semantics unchanged.

Another aspect that I would like to highlight in this patch that by attempting to preserve semantics of `rebuild_append_simple_rel_and_rte()` I have faced with ambiguity of handling of [RTE_VOID case](https://github.com/greenplum-db/gpdb/blob/6.10.1/src/backend/optimizer/plan/plangroupext.c#L2274-L2291). The main loop initializes RelOptInfo for RTE_VOID (former pulled up subquery) but doesn't save it to anywhere. I suppose it was overlooked and added the line that explicitly saves RelOptInfo into `root->simple_rel_array`. But I don't know how much it's correct. Anyway, the script below covers the RTE_VOID case and the presence of saving operation doesn't affect on behavior of query anyhow. I propose to determine fate of this line in current PR. @guofengrichard is author these lines of code and I would like to see some comments on this fragment.

## Script to reproduce

```
CREATE TABLE foo(type INTEGER, prod VARCHAR, quantity NUMERIC) 
DISTRIBUTED RANDOMLY
PARTITION BY RANGE (quantity) (                              
        partition "1" start (0) end (100)
);

SET optimizer TO off;

# failed query
SELECT type, prod, sum(quantity) s_quant
FROM (SELECT * FROM foo) t
GROUP BY GROUPING SETS((type), (prod));
```

## Affected versions

The buggy `rebuild_append_simple_rel_and_rte()` routine was introduced by commit 598f4b0ef7a00f4a629bfc2b6de42782b3cd538a and hence only GP6 releases are affected. 

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by hlinnaka** [comment]
@weinan003, this seems similar to the issue you fixed in commit f913cba22f2cc6b08a843439eef1fe1286bf53a9. Can you take a look, please?

File: src/test/regress/sql/partition_with_groupingsets.sql [hunk 1]
```diff
@@ -0,0 +1,44 @@
+--
+-- Drop existing table
+--
+DROP TABLE IF EXISTS foo;
+
+--
+-- Create new table foo
+--
+CREATE TABLE foo(type INTEGER, prod VARCHAR, quantity NUMERIC)
```
Comment (hlinnaka) [inline src/test/regress/sql/partition_with_groupingsets.sql:9]: Let's call this test table something else; there's like a dozen test in the test suite that use a table called "foo", and it's easy for them to clash if you try to run them in parallel. Or better yet, move this to `aggregate_with_groupingsets` test, which looks almost identical to this.

Comment (maksm90) [reply src/test/regress/sql/partition_with_groupingsets.sql]: Thanks for your note. Moved test case to `aggregate_with_groupingsets` test

**Comment by maksm90** [comment]
> @weinan003, this seems similar to the issue you fixed in commit [f913cba](https://github.com/greenplum-db/gpdb/commit/f913cba22f2cc6b08a843439eef1fe1286bf53a9). Can you take a look, please?

this commit doesn't fix current issue

**Comment by kainwen** [comment]
I have tested this pr for the issue: https://github.com/greenplum-db/gpdb/issues/10710

This pr should fix that issue also.  And all the cases in this pr pass.

@gaos1 @weinan003 Please help review this and get this in. Thanks.

**Comment by maksm90** [comment]
Hi, @kainwen 

> I have tested this pr for the issue: #10710
> 
> This pr should fix that issue also.

Yes, `cube` statement also assumes multiple canonical rollups in itself . So your pointed issue and this one are similar.

File: src/backend/optimizer/plan/plangroupext.c [hunk 2]
```diff
@@ -2256,23 +2255,56 @@ rebuild_append_simple_rel_and_rte(PlannerInfo *root,
...
 		if (rte->rtekind == RTE_VOID)
+		{
 			rel = makeNode(RelOptInfo);
+
+			/*
+			 * XXX: save empty RelOptInfo that corresponds to deleted RTE
+			 * (former pulled up subquery) in common array. This can be
+			 * incorrect decision.
+			 */
+			root->simple_rel_array[i] = rel;
```
Comment (darthunix) [inline src/backend/optimizer/plan/plangroupext.c:2306]: I am also confused about `RTE_VOID` - may be it is more correct to skip it in the manner we do with `RTE_JOIN`. In all other GPDB code `RTE_VOID` is either skipped or raise error/assert.

**Review by weinan003 — APPROVED** [review]
Thank you for contribute gpdb! I think this PR is fine to merge now.

**Pull request merged**